### PR TITLE
linker: non-recursive version of get_symbols

### PR DESCRIPTION
**Pull request by kroening**
This prevents a stack overflow in cases of very long dependency chains.

### CODE REVIEW

File: src/linking/remove_internal_symbols.cpp [hunk 1]
```diff
@@ -19,45 +19,44 @@ Author: Daniel Kroening
...
 #include "static_lifetime_init.h"
 
-void get_symbols_rec(
+static void get_symbols(
   const namespacet &ns,
   const symbolt &symbol,
   find_symbols_sett &dest)
 {
-  dest.insert(symbol.name);
+  std::vector<const symbolt *> working_set;
```
Comment (tautschnig) [inline src/linking/remove_internal_symbols.cpp:27]: It would have used a `forward_list`, but it's probably ok.

**Review by tautschnig — APPROVED** [review]

File: src/linking/remove_internal_symbols.cpp (hunk 1 above)
Comment (kroening) [reply src/linking/remove_internal_symbols.cpp:27]: I like the density of the vector, and the fact that it won't actually do memory allocation once the working set has reached some size.

Comment (kroening) [reply src/linking/remove_internal_symbols.cpp:27]: Quick experiment: 14.7s vs. 16.2s, for total goto-cc time on a larger .c file, in favour of the vector.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 0a255fd97e100878a82f8cb20589776431f9f175).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Pull request merged**